kbase/narrative
# SCT-2778: convert public data tab's refseq data source to use searchapi2 (was using legacy endpoint.)

# Replace usage of KBaseSearchEngine with searchapi2

The primary purpose is to replace usage of the KBaseSearchEngine api (the searchapi2/legacy endpoint) with the  searchapi2 api (the searchapi2/rpc endpoint) within the refseq data source of the public data tab of the data slide out.

The reason for this change is that the searchapi2/legacy endpoint suffers from problems, one of which broke this tool in CI, and it was deemed better to modify the ui to switch switch away from the api rather than fix the api. The searchapi2/rpc endpoint is the new search api for the new search back end.

The Narrative currently has two dependencies on the legacy search interface - the "NCBI RefSeq Genomes" (aka "refseq") widget in the Public tab of the data slide out, and the 'search' data source for the dynamic_dropdown.

In this PR the refseq widget is updated to run against search2, the dynamic_dropdown 'search' data source is removed, all search1 support removed, and search2 support added.

This change requires that search2 be available and the indexes up to date wherever the Narrative is deployed.

## Changes in this PR

The primary purpose of this set of changes is to have the refseq tool in the public data tab of the data slideout use the new searchapi2 api rather than the legacy searchapi2 api. This change is required by a regression in the legacy api which has broken the refseq tool. It was decided to just switch to the new api rather than fix the legacy api.

In the quest to add integration tests to cover the changes, I needed to enhance the integration test support. In addition, various changes were needed to support the tests, such as `test-id` attributes to support hooking onto otherwise anonymous DOM nodes.

When preparing the PR, all touched code triggers code quality alerts, so code changes were made to address those concerns, even if the code was not directly touched in the set of changes.

### Port public data tab's refseq tool to searchapi2/rpc from searchapi2/legacy

The main motivation for this PR is to switch the data tab's refseq tool from the searchapi2 legacy api to the searhapi2 new standard api. The changes to the existing 

Since the new api is JSONRPC 2.0, I needed to add a (cheap) JSONRPC 2.0 generic client.

Summary of changes
- added an id to public data source metadata definitions, to allow usage of `test-id` on fields generated metadata fields (allows selection of the fields for value testing)
- removed all legacy search support, including the configuration and client
- added JSONRPC 2.0 generic client.
- added minimal searchAPI2 client
- refactored narrative login, initially because it was touched as part of diagnosing test failure, secondarily due to code quality alerts triggered, and thirdly to address issues discovered when inspecting the code
- bootstrap dialog component: added onHide and onHidden to allow running code no matter how the dialog is closed (fixed latent bug in login)
- removed search support from dynamic dropdown; it was not used, and supplanted by general dynamic service support
- refactor data panel tabs to use direct clicks on tabs rather than delegated clicks; easier to test with
- removed searchDataSource (legacy), added search2DataSource
- minor changes to workspaceDataSource to address code quality
- new dependencies
    - selenium-standalone
    - @wdio/selenium-standlone-service
    - @wdio/browserstack-service
    - eslint
- config.json, config.json.tmpl: remove KBaseSearchEngine (legacy), add searchapi2.

### Integration test enhancments

- added test cases to existng narrative basic test
- aded tests for the refseq public data tool
- added support for selenium-standalone, browserstack services
- added support for many more test parameters
- run_tests.py: enable running against narratives at arbitrary url

### Integration tests for the refseq data tool

- added 5 new interation tests for the refseq data tool
- as noted above, enhanced the integration test tools in order to support running tests against more test services, any url, and with additional test parameters

### Various codebase fixes to enable tests

Several locations in both code files already changed as well as otherwise-untouched files involved in rendering surrounding markup were modified to support testing. Such changes were typically the addition of `data-test-id` attributes or roles.

### Various codebase fixes to address github code quality alerts

There were quite a few locations in all touched code which required primarily minor changes to address local linting issues as well as code quality issues reported at github.

### Misc

install_narrative.sh

replace code install with pip install (per Bill)
replace global bower usage with npx bower usage to use local bower

# JIRA Tickets
- https://kbase-jira.atlassian.net/browse/SCT-2778
- https://kbase-jira.atlassian.net/browse/SCT-2783
- [x] Added the Jira Ticket to the title of the PR 

# Testing Instructions
Integration tests have been added to exercise the several use cases, more can certainly be added. The tests should run against CI or narrative-dev. The test mechanism has been enhanced to support specifying the deploy environment to run against, a base host, the testing service, testing service auth, browser & version, os & version, screen size, and a few presets of these configurations, all as environment variables.
I still need to write that up, but I made sure that the plain "make test-integration" works if there is a token in the environment or in the token file.

- [x] Tests pass in GitHub Actions

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/truss.works/kbasetruss/development; it roughly does, much of that document applies to the KBase/Truss project
- [x] Any dependent changes have been merged and published in downstream modules (not applicable in this case so "any" is actually "none")
  - actually the IDE I use (VSC) has Flake8 linting and formatting built in; I modified just one Python file, which is not involved in the Narrative runtime - I hope that is enough.

# Updating Version and Release Notes (if applicable)

- [x] [Version has been bumped](https://semver.org/) for each release (not applicable in this case, but I did add a new "NEXT" release to the release notes.)
